gematria/datasets/exegesis_benchmark.cc

```diff
@@ -0,0 +1,325 @@
+
+  if (!MemoryMappings)
+    return llvm::make_error<StringError>(
+        errc::invalid_argument,
+        "Malformed basic block: no memory mappings array");
+
+  for (const auto &MemoryMappingValue : *MemoryMappings) {
+    const json::Object *MemoryMappingObject = MemoryMappingValue.getAsObject();
+
+    if (!MemoryMappingObject)
+      return llvm::make_error<StringError>(
+          errc::invalid_argument,
+          "Malformed memory mapping: memory mapping is not a JSON object");
+
+    std::optional<StringRef> MemoryMappingDefinitionName =
+        MemoryMappingObject->getString("Value");
+    std::optional<uintptr_t> MemoryMappingAddress =
+        MemoryMappingObject->getInteger("Address");
+
+    if (!MemoryMappingDefinitionName.has_value() ||
+        !MemoryMappingAddress.has_value())
+      return llvm::make_error<StringError>(
+          errc::invalid_argument,
+          "Malformed memory mapping: no name or address");
+
+    MemoryMapping MemMap;
+    MemMap.Address = *MemoryMappingAddress;
+    MemMap.MemoryValueName = MemoryMappingDefinitionName->str();
+    BenchCode.Key.MemoryMappings.push_back(MemMap);
+  }
+
+  return BenchCode;
+}
+
+Expected<unsigned> benchmarkBasicBlock(const BenchmarkCode &BenchCode,
+                                       const BenchmarkRunner &BenchRunner,
+                                       const LLVMState &State) {
+  std::unique_ptr<const SnippetRepetitor> SnipRepetitor =
+      SnippetRepetitor::Create(Benchmark::RepetitionModeE::MiddleHalfLoop,
+                               State, BenchCode.Key.LoopRegister);
+
+  SmallVector<Benchmark, 2> AllResults;
+
+  auto RC1 =
+      BenchRunner.getRunnableConfiguration(BenchCode, 5000, 0, *SnipRepetitor);
+  if (!RC1) return RC1.takeError();
+  auto RC2 =
+      BenchRunner.getRunnableConfiguration(BenchCode, 10000, 0, *SnipRepetitor);
+  if (!RC2) return RC2.takeError();
+
+  std::pair<Error, Benchmark> BenchmarkResult1OrErr =
+      BenchRunner.runConfiguration(std::move(*RC1), {});
+
+  if (std::get<0>(BenchmarkResult1OrErr))
+    return std::move(std::get<0>(BenchmarkResult1OrErr));
+
+  AllResults.push_back(std::move(std::get<1>(BenchmarkResult1OrErr)));
+
+  std::pair<Error, Benchmark> BenchmarkResult2OrErr =
+      BenchRunner.runConfiguration(std::move(*RC2), {});
+
+  if (std::get<0>(BenchmarkResult2OrErr))
+    return std::move(std::get<0>(BenchmarkResult2OrErr));
+
+  AllResults.push_back(std::move(std::get<1>(BenchmarkResult2OrErr)));
+
+  std::unique_ptr<ResultAggregator> ResultAgg =
+      ResultAggregator::CreateAggregator(
+          Benchmark::RepetitionModeE::MiddleHalfLoop);
+
+  Benchmark Result = std::move(AllResults[0]);
+
+  ResultAgg->AggregateResults(Result,
+                              ArrayRef<Benchmark>(AllResults).drop_front());
+
+  unsigned Throughput100 = static_cast<unsigned>(
+      round(Result.Measurements[0].PerSnippetValue * 100));
```

Comment:
I've switched to returning the raw snippet value for now. We might want to make this a configurable in the future to enable easy comparisons with BHive datasets.

Also, do we need to do anything on the ML side for this change? I guess fp32 shouldn't have any problems adapting to the smaller values?

Reply:
On the ML side, we should be fine with any scaling, as long as the scaling is consistent. That said, I'd really go for normalizing to a single repetition, so that we don't need to worry about the number of repetitions of each data set (I already got burned by using different normalizations/scaling factors in the past, and I don't want to repeat that :))

Reply:
Sounds good. The raw snippet value out of `llvm-exegesis` will already be normalized to a single repetition, so we'll be good there.